flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImpl.java

```diff
@@ -0,0 +1,194 @@
+	}
+
+	/**
+	 * This method wraps partition writers for externally managed partitions and introduces callbacks into the lifecycle
+	 * methods of the {@link ResultPartitionWriter}.
+	 */
+	@Override
+	public Collection<NotifyingResultPartitionWriter> createResultPartitionWriters(
+		JobID jobId,
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup) {
+
+		final Collection<? extends ResultPartitionWriter> resultPartitionWriters = backingShuffleEnvironment
+			.createResultPartitionWriters(taskName, executionAttemptID, resultPartitionDeploymentDescriptors, outputGroup, buffersGroup);
+
+		// determine which partitions are externally managed, and whose writers need to be wrapped
+		final Set<ResultPartitionID> externallyManagedPartitions = resultPartitionDeploymentDescriptors.stream()
+			.filter(ResultPartitionDeploymentDescriptor::isManagedExternally)
+			// TODO: The descriptor should be extended to also contain the ResultPartitionID
+			.map(descriptor -> new ResultPartitionID(descriptor.getPartitionId(), executionAttemptID))
+			.collect(Collectors.toSet());
+
+		// wrap partition writers to introduce callbacks for lifecycle events
+		final Collection<NotifyingResultPartitionWriter> wrappedResultPartitionWriters = new ArrayList<>(resultPartitionWriters.size());
+		for (ResultPartitionWriter resultPartitionWriter : resultPartitionWriters) {
+			final NotifyingResultPartitionWriter wrapper;
+			if (externallyManagedPartitions.contains(resultPartitionWriter.getPartitionId())) {
+				wrapper = new NotifyingResultPartitionWriter(
+					resultPartitionWriter,
+					partitionId -> notifyPartitionSetup(jobId, partitionId),
+					partitionId -> notifyPartitionFailed(jobId, partitionId),
+					partitionId -> notifyPartitionFinished(jobId, partitionId));
+			} else {
+				wrapper = new NotifyingResultPartitionWriter(
+					resultPartitionWriter,
+					NO_OP_NOTIFIER,
+					NO_OP_NOTIFIER,
+					NO_OP_NOTIFIER);
+			}
+			wrappedResultPartitionWriters.add(wrapper);
+		}
+
+		return wrappedResultPartitionWriters;
+	}
+
+	private void notifyPartitionSetup(JobID jobId, ResultPartitionID resultPartitionId) {
+		// if the job is marked inactive while a task is starting up it is possible that at this point we're adding
+		// a partition for an inactive job. This is fine; the task will go through the cancellation logic and fail the
+		// partition later.
+		inProgressPartitionTable.startTrackingPartition(jobId, resultPartitionId);
+	}
+
+	private void notifyPartitionFinished(JobID jobId, ResultPartitionID resultPartitionId) {
+		if (activeJobs.contains(jobId)) {
+			// keep this order to ensure #hasPartitionsOccupyingLocalResources returns true at all times
+			finishedPartitionTable.startTrackingPartition(jobId, resultPartitionId);
+			inProgressPartitionTable.stopTrackingPartition(jobId, resultPartitionId);
```

Comment:
`inProgressPartitionTable.stopTrackingPartition` would be called for both `if` and `else`, so extract it outside.